File: bubblesort.cpp
#include <iostream>

using namespace std;

const void doExchange(int *list,int index){
    int temp = list[index+1];
    list[index+1] = list[index];
    list[index] = temp;
}

void exchange(int *list, int length, int index=0, bool flag=true){
    bool isExchange = list[index]>list[index+1];
    if(isExchange){
        doExchange(list,index);
        flag = false;
    }
    if(index + 1 < length){
        exchange(list,length,index+1,flag);
    }else if(flag) return;
    else{
        exchange(list,length);
    }
}

void printSorted(int *list,int length){
    for(int i=0;i<length;i++) cout<<list[i]<<" ";
}

int main(){
    int a[10] = {2,4,32,45,3,1,5,9,6,7};
    exchange(a,10);
    printSorted(a,10);
    return 0;
}

File: bubblesort.rs

fn do_exchange(list: &mut Vec<usize>,index:usize) {
    let temp = list[index+1];
    list[index+1] = list[index];
    list[index] = temp;
}

fn sort(list: &mut Vec<usize>,index:usize,is_stable: bool ) {
    let length = list.len();
    let flag = list[index] < list[index+1];//is need changed
    if !flag{
        do_exchange(list,index);
    }
    let flag = flag && is_stable;
    if index + 2 < length{
        sort(list,index+1,flag);
    }else if flag {
        return;
    }else {
        sort(list,0,true)
    }
}

fn main() {
    let mut list : Vec<usize> = vec![2,3,1,4,6,8,0,5];
    sort(&mut list,0,true);
    println!("{:?}",list);
    let mut a,b :usize;
}